walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,69 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
+import java.util.Arrays;
+import java.util.Iterator;
+import com.google.sps.Pair;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      ArrayList<String> requiredPeople = new ArrayList<String>(request.getAttendees());
+      ArrayList<String> optionalPeople = new ArrayList<String>(request.getOptionalAttendees());
+
+      Collection<TimeRange> answer = maximizeOptionalAttendees(requiredPeople, optionalPeople, events, request.getDuration()).getValue1();
+      return answer;
+  }
+
+  private ArrayList<TimeRange> getbookedTimes(Collection<Event> events, Collection<String> people) {
+      ArrayList<TimeRange> bookedTimes = new ArrayList<>();
+      for(Event event: events) {
+          if(!Collections.disjoint(event.getAttendees(), people)){
+            bookedTimes.add(event.getWhen());
+          }
+      }
+      return bookedTimes;
+  }
+
+  private Collection<TimeRange> createAvailableRange(ArrayList<TimeRange> bookedTimes, long duration) {
+      Collection<TimeRange> availableTimes = new ArrayList<>();
+      Collections.sort(bookedTimes, TimeRange.ORDER_BY_START);
+      int nextSlot = TimeRange.START_OF_DAY;
+      for(TimeRange filled : bookedTimes) {
+          if(filled.start() - nextSlot >= duration) {
+              availableTimes.add(TimeRange.fromStartEnd(nextSlot, filled.start(), false));
+          }
+
+          if(filled.end() > nextSlot) {
+            nextSlot = filled.end();
+          }
+      }
+
+      if(TimeRange.END_OF_DAY - nextSlot >= duration) {
+          availableTimes.add(TimeRange.fromStartEnd(nextSlot, TimeRange.END_OF_DAY, true));
+      }
+      return availableTimes;
+  }
+
+  private Pair numberOfAttendees(ArrayList<String> people, Collection<Event> events, long duration){
+      ArrayList<TimeRange> bookedTime = getbookedTimes(events, people);
+      Collection<TimeRange> availableRange = createAvailableRange(bookedTime, duration);
+      Pair pair = new Pair(people.size(), availableRange);
+      return availableRange.isEmpty() ? new Pair(0, new ArrayList<TimeRange>()) : pair;
+  }
+
+  private Pair maximizeOptionalAttendees(ArrayList<String> requiredPeople, ArrayList<String> optionalPeople, Collection<Event> events, long duration){
+    if(optionalPeople.size()==0){
+        return numberOfAttendees(requiredPeople, events, duration);
+    }
+    ArrayList<String> temp2 = new ArrayList<String>(requiredPeople);
+    temp2.add(optionalPeople.get(optionalPeople.size()-1));
```

Comment:
This is a neat way to express that `optional` becomes `required`. However, we don't reduce the size of the optional set, and we also re-filter all of the events from scratch every time. We don't need both `required` and `events`. If we take the approach of incrementally filtering through the events, then we don't need to know any of the `required` participants, since `required` participants exist to just constrain the events.

Reply:
I'm conceptually having trouble separating the idea of people and events, because people can have multiple events, but we are trying to maximize number of people not number of slots/events